File: 백준/Gold/16562. 친구비/친구비.py
from sys import stdin

input = stdin.readline

n, m, k = map(int, input().split())

money = [0] + list(map(int, input().split()))

parents = [i for i in range(n + 1)]


def find_parent(x):
    global parents

    if parents[x] == x:
        return x

    parents[x] = find_parent(parents[x])
    return parents[x]


def union(x, y):
    global money, parents
    
    px, py = find_parent(x), find_parent(y)

    if py == px:
        return

    if money[py] < money[px]:
        py, px = px, py

    parents[py] = px

for a, b in [list(map(int, input().split())) for _ in range(m)]:
    union(a, b)

answer = 0
for i in range(1, n + 1):
    if parents[i] == i:
        answer += money[i]

print('Oh no' if answer > k else answer)

File: 백준/Gold/16562. 친구비/친구비.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

class Main{
    private static int n, money;
    private static int[] parents, costs;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());

        n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());
        money = Integer.parseInt(tokenizer.nextToken());
        
        parents = new int[n + 1];
        costs = new int[n + 1];
        tokenizer = new StringTokenizer(br.readLine());
        for(int i=1 ; i<=n ; i++){
            parents[i] = i;
            costs[i] = Integer.parseInt(tokenizer.nextToken());
        }

        while(m-- > 0){
            tokenizer = new StringTokenizer(br.readLine());
            int x, y;

            x = Integer.parseInt(tokenizer.nextToken());
            y = Integer.parseInt(tokenizer.nextToken());

            union(x, y);
        }

        Map<Integer, Boolean> visit = new HashMap<>();
        long answer = 0;
        for(int i=1 ; i<=n ; i++){
            int px = findParent(parents[i]);
            if(visit.containsKey(px))
                continue;

            answer += costs[px];
            visit.put(px, true);
        }

        System.out.println(answer == 0 || answer > money ? "Oh no" : answer);
    }
    
    private static int findParent(int x){
        if(x == parents[x])
            return x;
        return parents[x] = findParent(parents[x]);
    }

    private static void union(int x, int y){
        int px = findParent(x), py = findParent(y);

        if(px == py)
            return;

        if(costs[py] < costs[px]){
            int temp = py;
            py = px;
            px = temp;
        }

        parents[py] = px;
    }

}